draft-ietf-masque-connect-udp-listen.md

```diff
@@ -153,11 +170,103 @@ UDP Payload:
 : The unmodified UDP Payload of this proxied UDP packet (referred to as "data
 octets" in {{UDP}}).
 
+# Context ID {#contextid}
+
+Context ID values MUST always be greater than 0. A payload with value 0 MUST
+be dropped by the receiving party.
+
+An even context ID MUST be used by clients to send packets to the proxy, and an odd
+context ID MUST be used by the proxy to send packets to the client. A Client and
+a proxy MUST exhange COMPRESSION_ASSIGN packets for a given context ID to guarantee
+that all payloads are received and processed by the other party, see
+{{compression}}. They MAY pre-emptively use Context IDs not acknowledged by the
+other party yet, knowing that those packets MAY be lost since the client is not
+guaranteed to be ready to receive until it receives the acknowledgement
+COMPRESSION_ASSIGN.
+
+## Address Compression {#compression}
+
+If the client wishes to allocate a Context ID for uncompressed packets,
+it MUST exchange the COMPRESSION_ASSIGN capsule (see {{capsuleassignformat}})
+with the proxy with an unused Context ID defined in {{contextid}} with
+the IP Length set to zero.
+
+The client and the proxy MAY choose to compress the IP and port information
+per datagram for a given target against the Context ID.
+In such a case, the client or the proxy MUST send a COMPRESSION_ASSIGN capsule
+(see {{capsuleassignformat}}) with the target information (see
+{{targetmappingformat}}) it wishes to compress and the other party (proxy or
+client respectively) MUST echo back with either a  COMPRESSION_ASSIGN capsule
+if it accepts the compression request, or a  COMPRESSION_CLOSE with the context
+ID (see {{capsulecloseformat}}) if it doesn't wish to support  compression for
+the given Context ID (For example, due to \considerable memory requirements of
+establishing  a list of mappings per target per client). If the compression was
+rejected, the client and proxy MUST use an uncompressed context ID to exhange
+UDP payloads for the given target.
+
+
+### Compression Mapping {#mappings}
+
+If a COMPRESSION_ASSIGN is received by client or proxy with IP length non zero,
```

Comment:
This somewhat contradicts the previous paragraph because it ignores the fact that any endpoint can reject the compression requested by the peer. Perhaps:

When an endpoint receives a COMPRESSION_ASSIGN capsule with a non-zero IP length, it MUST decide whether to accept or reject the compression mapping:
* if it accepts the mapping, first the receiver MUST save the mapping from context ID to address and port. Second, the receiver MUST echo an identical COMPRESSION_ASSIGN capsule back to its peer.
* if it rejects the mapping, the receiver MUST respond by sending a COMPRESSION_CLOSE capsule with the context ID set to the one from the received COMPRESSION_ASSIGN capsule